apps/src/music/progress/ConditionsChecker.ts

```diff
@@ -17,21 +17,34 @@ export default class ConditionsChecker {
   }
 
   // Accumulate a satisfied condition.
-  addSatisfiedCondition(id: string) {
-    this.currentSatisfiedConditions.push(id);
+  addSatisfiedCondition(condition: KnownCondition) {
+    if (!this.isConditionSatisfied(condition)) {
+      this.currentSatisfiedConditions.push(condition);
+    }
+  }
+
+  private isConditionSatisfied(condition: KnownCondition) {
+    return this.currentSatisfiedConditions.some(
+      currentSatisfiedCondition =>
+        JSON.stringify(currentSatisfiedCondition) === JSON.stringify(condition)
+    );
   }
 
   // Check whether the current set of satisfied conditions satisfy the given
   // required conditions.
-  checkRequirementConditions(requiredConditions: string[]) {
+  checkRequirementConditions(requiredConditions: KnownCondition[]) {
```

Comment:
nit: since we're now using `KnownCondition` to reference both known and unknown conditions, maybe just `Condition` is a clearer name for the interface?

Reply:
Sounds good.